### PR TITLE
Update DataFusion architecture documentation

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes  https://github.com/apache/arrow-datafusion/issues/5501

# Rationale for this change
I am trying to make it easier to work on and use the DataFusion project

# What changes are included in this PR?

1. Update the architecture guide 

# Are these changes tested?

No (though  the links are verified by the changes in https://github.com/apache/arrow-datafusion/pull/6044) 

# Are there any user-facing changes?

Better docs

Here is an example:

Screenshot 2023-04-20 at 8 17 25 AM

### CODE REVIEW

**Review by comphead — APPROVED** [review]
lgtm, thanks

File: datafusion/core/src/lib.rs [hunk 1]
```diff
@@ -163,104 +168,238 @@
...
+//! 1. The [`LogicalPlan`] is checked and rewritten to enforce
+//! semantic rules, such as type coercion, by [`AnalyzerRule`]s
+//!
+//! 2. The [`LogicalPlan`] is rewritten by [`OptimizerRule`]s, such as
+//! projection and filter pushdown, to improve its efficiency.
+//!
+//! 3. The [`LogicalPlan`] is converted to an [`ExecutionPlan`] by a
+//! [`PhysicalPlanner`]
+//!
+//! 4. The [`ExecutionPlan`] is rewrittten by
```
Comment (Ted-Jiang) [inline datafusion/core/src/lib.rs:237]:
```suggestion
//! 4. The [`ExecutionPlan`] is rewritten by
```

File: datafusion/core/src/lib.rs [hunk 2]
```diff
@@ -163,104 +168,238 @@
...
+//!
+//! DataFusion includes several built in data sources for common use
+//! cases, and can be extended by implementing the [`TableProvider`]
+//! trait. A [`TableProvider`] provides information for planning and
+//! an [`ExecutionPlan`]s for execution.
 //!
-//! ### Physical planning
+//! 1. [`ListingTable`]: Reads data from Parquet, JSON, CSV, or AVRO
+//! files.  Supports single files or multiple files with HIVE style
+//! partitoning, optional compression, directly reading from remote
```
Comment (Ted-Jiang) [inline datafusion/core/src/lib.rs:266]:
```suggestion
//! partitioning, optional compression, directly reading from remote
```

**Review by Ted-Jiang — APPROVED** [review]
LGTM 👍

**Pull request merged**